libmuscle/python/libmuscle/instance.py

```diff
@@ -29,6 +31,62 @@
 _FInitCacheType = Dict[Tuple[str, Optional[int]], Message]
 
 
+class InstanceFlags(Flag):
+    """Enumeration of properties that an instance may have.
+
+    You may combine multiple flags using the bitwise OR operator `|`. For
+    example:
+
+    .. code-block:: python
+
+        from libmuscle import (
+                Instance, USES_CHECKPOINT_API, DONT_APPLY_OVERLAY)
+
+        ports = ...
+        flags = USES_CHECKPOINT_API | DONT_APPLY_OVERLAY
+        instance = Instance(ports, flags)
+    """
+
+    DONT_APPLY_OVERLAY = auto()
+    """Do not apply the received settings overlay during prereceive of F_INIT
+    messages. If you're going to use :meth:`Instance.receive_with_settings` on
+    your F_INIT ports, you need to set this flag when creating an
+    :class:`Instance`.
+
+    If you don't know what that means, do not specify this flag and everything
+    will be fine. If it turns out that you did need to specify the flag, MUSCLE3
+    will tell you about it in an error message and you can add it still.
+    """
+
+    USES_CHECKPOINT_API = auto()
+    """Indicate that this instance supports checkpointing.
+
+    You may not use any checkpointing API calls when this flag is not supplied.
+    """
+
+    HAS_NO_STATE_FOR_NEXT_USE = auto()
```

Comment:
KEEPS_NO_STATE_FOR_NEXT_USE? I think it may have a state because it can still iterate, but there may not be any gain to reusing the previous one, or it may even be worse. Naming these is tricky though...